File: macos/BolsaTerminal/Features/Bot/BotViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class BotViewModel {
    enum State {
        case idle, loading, loaded, error(String)
    }

    private struct ConfigPayload: Encodable {
        let enabled: Bool?
        let mode: String?
        let targetSymbols: [String]?
        let scanInterval: Double?
    }

    var state: State = .idle
    var status: BotStatus?
    var log: [BotLogEntry] = []
    var isStreaming = false

    var selectedMode: BotMode = .moderate
    var targetSymbolsText: String = ""
    var scanInterval: String = "60"

    private let sseClient = SSEClient()
    private let decoder = JSONDecoder()

    func load() async {
        state = .loading
        do {
            let fetchedStatus: BotStatus = try await APIClient.shared.request(Endpoints.botStatus)
            let fetchedLog: [BotLogEntry] = try await APIClient.shared.request(Endpoints.botLog(limit: 100))
            status = fetchedStatus
            log = fetchedLog
            selectedMode = fetchedStatus.config.mode
            scanInterval = String(Int(fetchedStatus.config.scanInterval))
            targetSymbolsText = fetchedStatus.config.targetSymbols.joined(separator: ", ")
            state = .loaded
            startStreaming()
        } catch {
            state = .error((error as? LocalizedError)?.errorDescription ?? error.localizedDescription)
        }
    }

    func startStreaming() {
        guard !isStreaming else { return }
        isStreaming = true
        sseClient.connect(endpoint: Endpoints.botStream) { [weak self] data in
            guard let self else { return }
            Task { @MainActor in
                guard let newEntries = try? self.decoder.decode([BotLogEntry].self, from: data) else { return }
                let existingIds = Set(self.log.map(\.id))
                let toInsert = newEntries.filter { !existingIds.contains($0.id) }
                guard !toInsert.isEmpty else { return }
                self.log = (toInsert + self.log)
                if self.log.count > 300 { self.log = Array(self.log.prefix(300)) }
            }
        } onComplete: { [weak self] _ in
            Task { @MainActor in self?.isStreaming = false }
        }
    }

    func stopStreaming() {
        sseClient.cancel()
        isStreaming = false
    }

    func toggleEnabled() async {
        guard let status else { return }
        await configure(enabled: !status.config.enabled)
    }

    func applyConfig() async {
        let symbols = targetSymbolsText
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespaces).uppercased() }
            .filter { !$0.isEmpty }
        await configure(mode: selectedMode, targetSymbols: symbols, scanInterval: Double(scanInterval) ?? 60)
    }

    private func configure(
        enabled: Bool? = nil,
        mode: BotMode? = nil,
        targetSymbols: [String]? = nil,
        scanInterval: Double? = nil
    ) async {
        let payload = ConfigPayload(
            enabled: enabled, mode: mode?.rawValue, targetSymbols: targetSymbols, scanInterval: scanInterval
        )
        do {
            let body = try JSONEncoder().encode(payload)
            status = try await APIClient.shared.request(Endpoints.botConfigure(body))
        } catch {
            state = .error((error as? LocalizedError)?.errorDescription ?? error.localizedDescription)
        }
    }
}

File: android/features/bot/src/main/java/com/bolsaterminal/features/bot/BotViewModel.kt
package com.bolsaterminal.features.bot

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bolsaterminal.core.common.Constants
import com.bolsaterminal.core.common.UiState
import com.bolsaterminal.core.model.BotLogEntry
import com.bolsaterminal.core.model.BotMode
import com.bolsaterminal.core.model.BotStatus
import com.bolsaterminal.core.network.SseClient
import com.bolsaterminal.domain.repository.BotRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import javax.inject.Inject

@HiltViewModel
class BotViewModel @Inject constructor(
    private val repository: BotRepository,
    private val sseClient: SseClient,
    private val json: Json,
) : ViewModel() {

    private val _state = MutableStateFlow<UiState<BotStatus>>(UiState.Idle)
    val state: StateFlow<UiState<BotStatus>> = _state.asStateFlow()

    private val _log = MutableStateFlow<List<BotLogEntry>>(emptyList())
    val log: StateFlow<List<BotLogEntry>> = _log.asStateFlow()

    var selectedMode by mutableStateOf(BotMode.Moderate)
    var targetSymbolsText by mutableStateOf("")
    var scanInterval by mutableStateOf("60")
    var isStreaming by mutableStateOf(false)
        private set

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            _state.value = UiState.Loading
            try {
                val status = repository.getStatus()
                _log.value = repository.getLog(100)
                selectedMode = status.config.mode
                scanInterval = status.config.scanInterval.toInt().toString()
                targetSymbolsText = status.config.targetSymbols.joinToString(", ")
                _state.value = UiState.Loaded(status)
                startStreaming()
            } catch (e: Exception) {
                _state.value = UiState.Error(e.message.orEmpty())
            }
        }
    }

    private fun startStreaming() {
        if (isStreaming) return
        isStreaming = true
        sseClient.connect(
            path = Constants.SSE_BOT_STREAM_PATH,
            onEvent = { raw ->
                try {
                    val newEntries = json.decodeFromString<List<BotLogEntry>>(raw)
                    val existingIds = _log.value.map { it.id }.toSet()
                    val toInsert = newEntries.filter { it.id !in existingIds }
                    if (toInsert.isNotEmpty()) {
                        _log.value = (toInsert + _log.value).take(300)
                    }
                } catch (_: Exception) {
                    // Ping comments and malformed frames are ignored — best-effort stream.
                }
            },
            onComplete = { isStreaming = false },
        )
    }

    override fun onCleared() {
        super.onCleared()
        sseClient.cancel()
    }

    fun toggleEnabled() {
        val current = (state.value as? UiState.Loaded)?.data ?: return
        configure(enabled = !current.config.enabled)
    }

    fun applyConfig() {
        val symbols = targetSymbolsText.split(",").map { it.trim().uppercase() }.filter { it.isNotEmpty() }
        configure(mode = selectedMode.name.lowercase(), targetSymbols = symbols, scanInterval = scanInterval.toDoubleOrNull() ?: 60.0)
    }

    private fun configure(
        enabled: Boolean? = null,
        mode: String? = null,
        targetSymbols: List<String>? = null,
        scanInterval: Double? = null,
    ) {
        viewModelScope.launch {
            try {
                _state.value = UiState.Loaded(repository.configure(enabled, mode, targetSymbols, scanInterval))
            } catch (e: Exception) {
                _state.value = UiState.Error(e.message.orEmpty())
            }
        }
    }
}
